workflow/artifacts/gcs/gcs.go

```diff
@@ -0,0 +1,129 @@
+package gcs
+
+import (
+	"cloud.google.com/go/storage"
+	"context"
+	"errors"
+	argoErrors "github.com/argoproj/argo/errors"
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"os"
+)
+
+type GCSArtifactDriver struct {
+	Context context.Context
+}
+
+func (gcsDriver *GCSArtifactDriver) newGcsClient() (client *storage.Client, err error) {
+	gcsDriver.Context = context.Background()
+	client, err = storage.NewClient(gcsDriver.Context)
+	if err != nil {
+		return nil, argoErrors.InternalWrapError(err)
+	}
+	return
+
+}
+
+func (gcsDriver *GCSArtifactDriver) saveToFile(inputArtifact *wfv1.Artifact, filePath string) error {
+
+	log.Infof("Loading from GCS (gs://%s/%s) to %s",
+		inputArtifact.GCS.Bucket, inputArtifact.GCS.Key, filePath)
+
+	stat, err := os.Stat(filePath)
+	if err == nil {
+		if stat.IsDir() {
+			return errors.New("output artifact path is a directory")
+		}
+	}
+
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+	}
+
+	outputFile, err := os.OpenFile(filePath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+
+	gcsClient, err := gcsDriver.newGcsClient()
+	if err != nil {
+		return err
+	}
+
+	bucket := gcsClient.Bucket(inputArtifact.GCS.Bucket)
```

Comment:
I'm assuming we don't have to do any empty-string checks due to how the serialization works (else you should add them).